tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+                            &lut,
+                        )
+                    },
+                    || {
+                        let (shifted_blocks, mut block_states) =
+                            self.compute_shifted_blocks_and_block_states(blocks);
+                        let _ = block_states.pop().unwrap();
+                        (shifted_blocks, block_states)
+                    },
+                );
+
+                output_flag = Some(block);
+                (shifted_blocks, block_states)
+            }
+            OutputFlag::Carry => {
+                let (shifted_blocks, mut block_states) =
+                    self.compute_shifted_blocks_and_block_states(blocks);
+                let last_block_state = block_states.pop().unwrap();
+                output_flag = Some(last_block_state);
+                (shifted_blocks, block_states)
+            }
+        };
+
+        // Second step
+        let (mut prepared_blocks, mut groupings_pgns) = {
+            // This stores, the LUTs that given a cum sum block in the first grouping
+            // tells if a carry is generated or not
+            let first_grouping_inner_propagation_luts = (0..grouping_size - 1)
+                .map(|index| {
+                    self.key.generate_lookup_table(|propa_cum_sum_block| {
+                        let carry = propa_cum_sum_block & (1 << index);
+                        if carry != 0 {
+                            2 // Generates
+                        } else {
+                            0 // Nothing
+                        }
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            // This stores, the LUTs that given a cum sum in non first grouping
+            // tells if a carry is generated or propagated or neither of these
+            let other_groupings_inner_propagation_luts = (0..grouping_size)
+                .map(|index| {
+                    self.key.generate_lookup_table(|propa_cum_sum_block| {
+                        let mask = (2 << index) - 1;
+                        if propa_cum_sum_block >= (2 << index) {
+                            2 // Generates
+                        } else if (propa_cum_sum_block & mask) == mask {
+                            1 // Propagate
+                        } else {
+                            0
+                        }
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            // This stores the LUT that outputs the propagation result of the first grouping
+            let first_grouping_outer_propagation_lut = self.key.generate_lookup_table(|block| {
+                // Check if the last bit of the block is set
+                (block >> (num_bits_in_block - 1)) & 1
+            });
+
+            // This stores the LUTs that output the propagation result of the other groupings
+            let grouping_chunk_pgn_luts = if use_sequential_algorithm_to_resolved_grouping_carries {
```

Comment:
I'm not sure having two differente algorithms in the same function is a good idea here ? could be some sub functions maybe ?

Reply:
Yes that can probably be done

Reply:
if it makes sense, but I think it makes quite a bit of state to keep in ones head, so I let you be the judge